### ISSUE TITLE
Server shouldn't be able to establish routes to itself

### ISSUE THREAD

**Issue by pires** [nats-io/gnatsd#175]
How to reproduce with latest `master`:
```
$ ./gnatsd --cluster nats://localhost:6222 --routes nats://localhost:6222 -DV
```

Here's what I see:
```
[35433] 2016/01/20 15:23:07.177801 [INF] Starting gnatsd version 0.7.3
[35433] 2016/01/20 15:23:07.177919 [INF] Listening for route connections on localhost:6222
[35433] 2016/01/20 15:23:07.180640 [INF] Listening for client connections on 0.0.0.0:4222
[35433] 2016/01/20 15:23:07.180729 [INF] gnatsd is ready
[35433] 2016/01/20 15:23:07.180766 [DBG] Trying to connect to route on localhost:6222
[35433] 2016/01/20 15:23:07.182566 [DBG] 127.0.0.1:6222 - rid:1 - Route connection created
[35433] 2016/01/20 15:23:07.182582 [DBG] 127.0.0.1:6222 - rid:1 - Route connect msg sent
[35433] 2016/01/20 15:23:07.182666 [DBG] 127.0.0.1:54390 - rid:2 - Route connection created
[35433] 2016/01/20 15:23:07.182864 [TRC] 127.0.0.1:54390 - rid:2 - ->> [CONNECT {"verbose":false,"pedantic":false,"tls_required":false,"name":"8313e507c8dee5b540e80ecbbd2ae556"}]
[35433] 2016/01/20 15:23:07.182883 [DBG] 127.0.0.1:6222 - rid:1 - Registering remote route "8313e507c8dee5b540e80ecbbd2ae556"
[35433] 2016/01/20 15:23:07.182891 [DBG] 127.0.0.1:6222 - rid:1 - Route sent local subscriptions
[35433] 2016/01/20 15:23:07.182939 [DBG] 127.0.0.1:54390 - rid:2 - Detected duplicate remote route "8313e507c8dee5b540e80ecbbd2ae556"
[35433] 2016/01/20 15:23:07.182945 [DBG] 127.0.0.1:54390 - rid:2 - Router connection closed
[35433] 2016/01/20 15:23:07.183000 [DBG] 127.0.0.1:6222 - rid:1 - Router connection closed
[35433] 2016/01/20 15:23:07.183035 [DBG] Attempting reconnect for solicited route "nats://localhost:6222"
[35433] 2016/01/20 15:23:08.183622 [DBG] Trying to connect to route on localhost:6222
[35433] 2016/01/20 15:23:08.186011 [DBG] 127.0.0.1:6222 - rid:3 - Route connection created
[35433] 2016/01/20 15:23:08.186029 [DBG] 127.0.0.1:6222 - rid:3 - Route connect msg sent
[35433] 2016/01/20 15:23:08.186041 [DBG] 127.0.0.1:54392 - rid:4 - Route connection created
[35433] 2016/01/20 15:23:08.186186 [TRC] 127.0.0.1:54392 - rid:4 - ->> [CONNECT {"verbose":false,"pedantic":false,"tls_required":false,"name":"8313e507c8dee5b540e80ecbbd2ae556"}]
[35433] 2016/01/20 15:23:08.186212 [DBG] 127.0.0.1:6222 - rid:3 - Registering remote route "8313e507c8dee5b540e80ecbbd2ae556"
[35433] 2016/01/20 15:23:08.186219 [DBG] 127.0.0.1:6222 - rid:3 - Route sent local subscriptions
[35433] 2016/01/20 15:23:08.186251 [DBG] 127.0.0.1:54392 - rid:4 - Detected duplicate remote route "8313e507c8dee5b540e80ecbbd2ae556"
[35433] 2016/01/20 15:23:08.186256 [DBG] 127.0.0.1:54392 - rid:4 - Router connection closed
[35433] 2016/01/20 15:23:08.186302 [DBG] 127.0.0.1:6222 - rid:3 - Router connection closed
[35433] 2016/01/20 15:23:08.186331 [DBG] Attempting reconnect for solicited route "nats://localhost:6222"
[35433] 2016/01/20 15:23:09.189305 [DBG] Trying to connect to route on localhost:6222
[35433] 2016/01/20 15:23:09.191021 [DBG] 127.0.0.1:6222 - rid:5 - Route connection created
[35433] 2016/01/20 15:23:09.191038 [DBG] 127.0.0.1:6222 - rid:5 - Route connect msg sent
[35433] 2016/01/20 15:23:09.191110 [DBG] 127.0.0.1:54394 - rid:6 - Route connection created
[35433] 2016/01/20 15:23:09.191214 [TRC] 127.0.0.1:54394 - rid:6 - ->> [CONNECT {"verbose":false,"pedantic":false,"tls_required":false,"name":"8313e507c8dee5b540e80ecbbd2ae556"}]
[35433] 2016/01/20 15:23:09.191230 [DBG] 127.0.0.1:6222 - rid:5 - Registering remote route "8313e507c8dee5b540e80ecbbd2ae556"
[35433] 2016/01/20 15:23:09.191241 [DBG] 127.0.0.1:6222 - rid:5 - Route sent local subscriptions
[35433] 2016/01/20 15:23:09.191253 [DBG] 127.0.0.1:54394 - rid:6 - Detected duplicate remote route "8313e507c8dee5b540e80ecbbd2ae556"
[35433] 2016/01/20 15:23:09.191258 [DBG] 127.0.0.1:54394 - rid:6 - Router connection closed
[35433] 2016/01/20 15:23:09.191302 [DBG] 127.0.0.1:6222 - rid:5 - Router connection closed
[35433] 2016/01/20 15:23:09.191334 [DBG] Attempting reconnect for solicited route "nats://localhost:6222"
```

**Comment by kozlovic** [comment]
@pires: Just to confirm. Your expectation is that the server would fail to start and report an invalid configuration or something else?

**Comment by derekcollison** [comment]
No, it should simply ignore. Similar to when servers A and B are configured to connect to each other, we detect duplicate connections and simply drop the redundant one.

**Comment by kozlovic** [comment]
Well, this is what's happening here. It detects and drop, but try again ;-). But yes, I understand, detects first so we don't even try.

**Comment by pires** [comment]
> No, it should simply ignore.

Exactly.

**Comment by kozlovic** [comment]
In multi-home servers, the host may resolve do different IPs. So I should get the actual IP use for the route listen spec, and verify that none of the resolved IPs for the "routes" host resolve to that IP. Right?

**Comment by pires** [comment]
What happens when you're listening on `0.0.0.0`?

I'd say an instance has a list of IPs it is listening on (`0.0.0.0` translates to all interfaces so one would need to know all available interface IP addresses) and ignores routes to any of those.

**Comment by pires** [comment]
Also, one problem I've noticed when running inside a Docker container (without `--net host` flag specified) is that NATS only knows the container internal IP, thus adding more complexity to what I suggested above.

Oh, and when running NATS on a container, it knows only the container internal IP and uses that when establishing a route, rendering it invalid.
```
[1] 2016/02/10 15:19:53.617147 [DBG] 172.17.42.1:36918 - rid:19 - Registering remote route "8e1dfc9250e9c867938916133334b2fa"
[1] 2016/02/10 15:19:53.617163 [DBG] 172.17.42.1:36918 - rid:19 - Route sent local subscriptions
[1] 2016/02/10 15:19:53.618411 [DBG] 172.30.5.140:6222 - rid:18 - Detected duplicate remote route "8e1dfc9250e9c867938916133334b2fa"
[1] 2016/02/10 15:19:53.618418 [DBG] 172.30.5.140:6222 - rid:18 - Router connection closed
[1] 2016/02/10 15:19:53.618450 [DBG] Not attempting reconnect for solicited route, already connected to "8e1dfc9250e9c867938916133334b2fa"
[1] 2016/02/10 15:19:53.618659 [DBG] 172.17.42.1:36918 - rid:19 - Router connection closed
[1] 2016/02/10 15:19:53.618691 [DBG] Attempting reconnect for solicited route "nats-route://ruser:T0pS3cr3t@172.30.5.140:6222"```

`172.17.42.1` is Docker networking and not routable from outside the host where this container is running.

**Comment by kozlovic** [comment]
Again, all that makes the idea of letting the route connect, and then detect that its from the server itself, and close the connection and not try to reconnect. The thing that is not nice about this is that quite a bit happens during the route create (sending all known routes, subscription interest, etc..), all for nothing.

**Comment by ColinSullivan1** [comment]
Paulo,

You're good at bringing up corner cases!

While there are a few ways to solve this I do think Ivan's idea (detect
oneself after connecting) is simplest and introduces the least amount of
complexity.  It also seems robust enough to handle complex cloud
environments...  my $.02.

Colin

On Wed, Feb 10, 2016 at 12:40 PM, Ivan Kozlovic <notifications@github.com>
wrote:

> Again, all that makes the idea of letting the route connect, and then
> detect that its from the server itself, and close the connection and not
> try to reconnect. The thing that is not nice about this is that quite a bit
> happens during the route create (sending all known routes, subscription
> interest, etc..), all for nothing.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nats-io/gnatsd/issues/175#issuecomment-182544536>.
>



-- 
*Colin Sullivan* | Principal Engineer
colin.sullivan@apcera.com <brian@apcera.com>  | @ColinSullivan01

**Comment by kozlovic** [comment]
Do I understand correctly that the output above corresponds to the case where `-routes` points to itself, right? I see that the "duplicate" route is detected, even though the ip don't match.

If so, that would be great because I am having a hard time detecting with certainty that we would be connecting to ourselves. I may go with the approach I've described before (detect that remoteID is server ID) and see if all tests pass. Comments?

**Comment by kozlovic** [comment]
@ColinSullivan1 Our updates crossed path. Good to see that the idea is not too unreasonable.

**Comment by pires** [comment]
> detect that remoteID is server ID

This should do it so I agree the connection may be established, invalidated and never tried again. Is this what you're proposing @kozlovic?

**Comment by pires** [comment]
Btw, with orchestration tools like Kubernetes this doesn't happen because it (the tool) owns the networking space and not Docker (overlay) so the container IP address is reachable from other containers (_pods_ in Kubernetes lingo).

**Comment by kozlovic** [comment]
Yes, that's what I was thinking. Experimenting.

**Comment by kozlovic** [comment]
Closed with #202.
